File: questions/11.ContainerWithMostWater.js
/**
 * @param {number[]} height
 * @return {number}
 */
const maxArea = function (height) {
  let result = 0
  let left = 0
  let right = height.length - 1

  while (left < right) {
    const w = right - left
    const h = Math.min(height[left], height[right])

    const area = w * h
    result = Math.max(result, area)

    if (height[left] < height[right]) {
      left++
      continue
    }
    right--
  }

  return result
}

console.log(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))

File: leetcode/11.ContainerWithMostWater.go
package leetcode

import "math"

func maxArea(height []int) int {
	result := 0
	left := 0
	right := len(height) - 1

	for left < right {
		area := (right - left) * int(math.Min(float64(height[left]), float64(height[right])))
		result = int(math.Max(float64(result), float64(area)))

		if height[left] < height[right] {
			left++
			continue
		}

		right--
	}

	return result
}
